File: kernel/sidecar/models/migration.py
"""Versioned migration runner for kairo.db (SQLite).

This is the Python-side mirror of ``overlay/src-tauri/src/migration.rs``.
Both must agree on SCHEMA_VERSION_CURRENT and migration semantics.

Design principles:
  - Schema version lives in ``PRAGMA user_version`` (built-in SQLite field).
  - Migrations are forward-only and idempotent (safe to re-run).
  - Each migration runs in a single transaction: ROLLBACK on failure.
  - A backup copy is created before any migration: ``kairo.db.bak.v<N>.<ts>``.
  - The backup is preserved even on success (user safety net).
  - Qdrant vector store versioning lives in a companion ``meta.json``:
    ``{"embedding_dim": N, "model": "...", "schema_version": N}``.
    If model/dim changes, a rebuild is triggered from SQLite chunk text.
  - A migration that fails leaves user_version unchanged and the original
    store fully usable.
"""

from __future__ import annotations

import json
import logging
import os
import pathlib
import shutil
import sqlite3
import time
from dataclasses import dataclass
from typing import Callable, List, Optional

log = logging.getLogger("kairo.migration")

# ── Version constants ────────────────────────────────────────────────────────

#: Current application schema version. Increment whenever a new migration is added.
#: Must match the Rust constant in ``migration.rs``.
SCHEMA_VERSION_CURRENT: int = 2

#: Default embedding model name written to Qdrant meta.json.
#: Matches the fallback _HashEmbedder used when sentence_transformers is absent.
DEFAULT_EMBEDDING_MODEL: str = "hash-256"

#: Default embedding dimension for the _HashEmbedder.
DEFAULT_EMBEDDING_DIM: int = 256


# ── Migration definition ─────────────────────────────────────────────────────

@dataclass
class Migration:
    """A single forward migration step."""

    to_version: int
    """Target schema version after this migration completes."""

    description: str
    """Human-readable description for logging and audit trail."""

    statements: List[str]
    """SQL statements to execute. Each is idempotent (uses IF NOT EXISTS, etc.)."""


def _col_exists(conn: sqlite3.Connection, table: str, column: str) -> bool:
    """Return True if `column` exists in `table`."""
    cur = conn.execute(f"PRAGMA table_info({table})")
    return any(row[1] == column for row in cur.fetchall())


def _make_statements_v0_to_v1() -> List[str]:
    """Build migration statements for v0 → v1.

    Uses Python-side column-existence check so the SQL itself stays idempotent.
    """
    return [
        # Audit log table (always safe — IF NOT EXISTS)
        """CREATE TABLE IF NOT EXISTS schema_migration_log (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            applied_at INTEGER NOT NULL,
            from_version INTEGER NOT NULL,
            to_version INTEGER NOT NULL,
            description TEXT NOT NULL
        )""",
        # documents.use_visual_grading — checked at runtime before issuing
        # (handled in _run_statement_idempotent via 'duplicate column name' catch)
        "ALTER TABLE documents ADD COLUMN use_visual_grading INTEGER NOT NULL DEFAULT 0",
    ]


def _make_statements_v1_to_v2() -> List[str]:
    return [
        """CREATE TABLE IF NOT EXISTS schema_migration_log (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            applied_at INTEGER NOT NULL,
            from_version INTEGER NOT NULL,
            to_version INTEGER NOT NULL,
            description TEXT NOT NULL
        )""",
        "ALTER TABLE pages ADD COLUMN dpi REAL NOT NULL DEFAULT 150.0",
        "ALTER TABLE extractions ADD COLUMN pack_name TEXT NOT NULL DEFAULT 'generic'",
    ]


#: All migrations in ascending version order.
MIGRATIONS: List[Migration] = [
    Migration(
        to_version=1,
        description="Establish schema_version tracking; add documents.use_visual_grading",
        statements=_make_statements_v0_to_v1(),
    ),
    Migration(
        to_version=2,
        description="Add pages.dpi; add extractions.pack_name",
        statements=_make_statements_v1_to_v2(),
    ),
]


# ── Public API ───────────────────────────────────────────────────────────────

class MigrationError(RuntimeError):
    """Raised when a migration fails. The database is left unchanged."""

    def __init__(self, message: str, to_version: Optional[int] = None):
        super().__init__(message)
        self.to_version = to_version


class VersionAheadError(MigrationError):
    """Raised when the database schema version is newer than the app."""

    def __init__(self, db_version: int, app_version: int):
        super().__init__(
            f"Database schema v{db_version} is newer than this app (v{app_version}). "
            f"Upgrade the app to open this database.",
        )
        self.db_version = db_version
        self.app_version = app_version


def run_migrations(db_path: pathlib.Path) -> None:
    """Run all pending migrations on the database at ``db_path``.

    Raises:
        MigrationError: if a migration SQL fails. Database is unchanged.
        VersionAheadError: if db schema is newer than app.
    """
    if not db_path.exists():
        log.debug("Database does not exist yet at %s — skipping migration.", db_path)
        return

    conn = sqlite3.connect(str(db_path))
    conn.execute("PRAGMA journal_mode = WAL")
    conn.execute("PRAGMA foreign_keys = ON")

    try:
        current_version = _get_user_version(conn)

        if current_version == SCHEMA_VERSION_CURRENT:
            log.debug("Schema is up to date (v%d). No migrations needed.", SCHEMA_VERSION_CURRENT)
            return

        if current_version > SCHEMA_VERSION_CURRENT:
            raise VersionAheadError(current_version, SCHEMA_VERSION_CURRENT)

        log.info(
            "Database is at v%d, app requires v%d. Running %d migration(s).",
            current_version,
            SCHEMA_VERSION_CURRENT,
            SCHEMA_VERSION_CURRENT - current_version,
        )

        # Create backup BEFORE touching anything.
        _create_backup(db_path, current_version)

        pending = [m for m in MIGRATIONS if m.to_version > current_version]
        for migration in pending:
            _run_single_migration(conn, migration, current_version)
            current_version = migration.to_version  # track for audit log

        log.info("All migrations complete. Schema is now v%d.", SCHEMA_VERSION_CURRENT)

    finally:
        conn.close()


def qdrant_needs_rebuild(
    qdrant_dir: pathlib.Path,
    expected_dim: int = DEFAULT_EMBEDDING_DIM,
    expected_model: str = DEFAULT_EMBEDDING_MODEL,
) -> bool:
    """Return True if the Qdrant vector store needs a rebuild.

    A rebuild is needed when:
      - ``meta.json`` exists but records a different ``embedding_dim`` or ``model``.
      - ``meta.json`` is corrupt/unreadable.

    A rebuild is NOT needed when:
      - ``meta.json`` does not exist (fresh store — will be created on first use).
      - ``meta.json`` records matching dim and model.
    """
    meta_path = qdrant_dir / "meta.json"
    if not meta_path.exists():
        return False  # Fresh store — no rebuild needed.

    try:
        meta = json.loads(meta_path.read_text(encoding="utf-8"))
        dim_ok = meta.get("embedding_dim") == expected_dim
        model_ok = meta.get("model") == expected_model
        return not (dim_ok and model_ok)
    except (json.JSONDecodeError, OSError) as e:
        log.warning("Could not read Qdrant meta.json: %s — assuming rebuild needed.", e)
        return True


def write_qdrant_meta(
    qdrant_dir: pathlib.Path,
    dim: int,
    model: str,
    schema_version: int = SCHEMA_VERSION_CURRENT,
) -> None:
    """Write a new Qdrant meta.json after a successful rebuild."""
    meta = {
        "embedding_dim": dim,
        "model": model,
        "schema_version": schema_version,
    }
    meta_path = qdrant_dir / "meta.json"
    meta_path.write_text(json.dumps(meta, indent=2), encoding="utf-8")
    log.info("Qdrant meta.json written: dim=%d, model=%s, schema_version=%d", dim, model, schema_version)


# ── Internal helpers ─────────────────────────────────────────────────────────

def _get_user_version(conn: sqlite3.Connection) -> int:
    row = conn.execute("PRAGMA user_version").fetchone()
    return int(row[0]) if row else 0


def _set_user_version(conn: sqlite3.Connection, version: int) -> None:
    conn.execute(f"PRAGMA user_version = {version}")


def _create_backup(db_path: pathlib.Path, old_version: int) -> pathlib.Path:
    ts = int(time.time())
    bak_name = f"{db_path.name}.bak.v{old_version}.{ts}"
    bak_path = db_path.parent / bak_name
    shutil.copy2(str(db_path), str(bak_path))
    log.info("Backup created at %s", bak_path)
    return bak_path


def _run_single_migration(
    conn: sqlite3.Connection,
    migration: Migration,
    from_version: int,
) -> None:
    """Execute one migration inside a transaction. Raises MigrationError on failure."""
    log.info(
        "Applying migration → v%d: %s",
        migration.to_version,
        migration.description,
    )

    try:
        conn.execute("BEGIN IMMEDIATE")

        for stmt in migration.statements:
            _run_statement_idempotent(conn, stmt)

        # Audit log (best-effort; table may not exist in first v0→v1 run before CREATE).
        ts = int(time.time())
        try:
            conn.execute(
                "INSERT INTO schema_migration_log "
                "(applied_at, from_version, to_version, description) VALUES (?, ?, ?, ?)",
                (ts, from_version, migration.to_version, migration.description),
            )
        except sqlite3.OperationalError:
            pass  # Audit log table didn't exist yet — created in this migration.

        # Set user_version within the same transaction.
        conn.execute(f"PRAGMA user_version = {migration.to_version}")

        conn.execute("COMMIT")
        log.info("✓ Migration to v%d succeeded.", migration.to_version)

    except Exception as exc:
        try:
            conn.execute("ROLLBACK")
        except Exception:
            pass
        raise MigrationError(
            f"Migration to v{migration.to_version} failed (rolled back): {exc}",
            to_version=migration.to_version,
        ) from exc


def _run_statement_idempotent(conn: sqlite3.Connection, sql: str) -> None:
    """Execute a SQL statement, treating 'duplicate column name' as a no-op."""
    try:
        conn.execute(sql)
    except sqlite3.OperationalError as e:
        msg = str(e).lower()
        if "duplicate column name" in msg or "already exists" in msg:
            log.debug("Idempotent no-op (column already exists): %s", e)
        else:
            raise

File: overlay/src-tauri/src/migration.rs
//! Versioned migration runner for kairo.db (SQLite).
//!
//! # Design
//!
//! - Schema version is stored in SQLite's built-in `PRAGMA user_version`.
//! - Migrations are forward-only and idempotent: running migration N twice
//!   produces the same schema as running it once.
//! - Each migration runs in a single transaction. On failure: ROLLBACK → the
//!   original schema and user_version are preserved.
//! - Before migrating, a backup copy is created: `kairo.db.bak.<old_version>.<ts>`.
//!   This backup is preserved even if the migration succeeds.
//! - Qdrant vector store versioning: a companion `meta.json` records
//!   `{"embedding_dim": N, "model": "...", "schema_version": N}`.
//!   If the model or dim changes, a rebuild is triggered from SQLite chunk text.

use rusqlite::{Connection, Result as SqlResult};
use std::path::{Path, PathBuf};
use std::time::{SystemTime, UNIX_EPOCH};

/// Current application schema version.
/// Increment this constant whenever a new migration is added.
pub const SCHEMA_VERSION_CURRENT: u32 = 2;

/// A single database migration step.
struct Migration {
    /// Target version after this migration runs.
    to_version: u32,
    /// Description for logging.
    description: &'static str,
    /// The SQL statements to execute (each is a separate statement).
    /// Statements must be idempotent (use IF NOT EXISTS, IF NOT EXISTS column checks, etc.)
    statements: &'static [&'static str],
}

/// All migrations in order. Each must be idempotent.
static MIGRATIONS: &[Migration] = &[
    Migration {
        to_version: 1,
        description: "Establish schema_version tracking; add documents.use_visual_grading",
        statements: &[
            // Create schema_migration_log table for audit trail.
            "CREATE TABLE IF NOT EXISTS schema_migration_log (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                applied_at INTEGER NOT NULL,
                from_version INTEGER NOT NULL,
                to_version INTEGER NOT NULL,
                description TEXT NOT NULL
            )",
            // Add use_visual_grading to documents if not present.
            // SQLite doesn't support IF NOT EXISTS on ALTER TABLE ADD COLUMN directly;
            // we use a try/ignore pattern by wrapping in a SELECT that checks the column.
            // The actual idempotency check is handled by the runner (see run_statement_idempotent).
            "ALTER TABLE documents ADD COLUMN use_visual_grading INTEGER NOT NULL DEFAULT 0",
        ],
    },
    Migration {
        to_version: 2,
        description: "Add pages.dpi; add extractions.pack_name; create schema_migration_log if missing",
        statements: &[
            "CREATE TABLE IF NOT EXISTS schema_migration_log (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                applied_at INTEGER NOT NULL,
                from_version INTEGER NOT NULL,
                to_version INTEGER NOT NULL,
                description TEXT NOT NULL
            )",
            "ALTER TABLE pages ADD COLUMN dpi REAL NOT NULL DEFAULT 150.0",
            "ALTER TABLE extractions ADD COLUMN pack_name TEXT NOT NULL DEFAULT 'generic'",
        ],
    },
];

/// Error type for migration failures.
#[derive(Debug)]
pub enum MigrationError {
    /// The database file could not be opened or backed up.
    Io(std::io::Error),
    /// A SQL statement failed. The migration was rolled back.
    /// `u32` is the target version that failed; `String` is the rusqlite error.
    SqlFailed(u32, String),
    /// The database schema version is ahead of SCHEMA_VERSION_CURRENT.
    /// This means a newer app wrote this database; we refuse to downgrade.
    VersionAhead { db_version: u32, app_version: u32 },
}

impl std::fmt::Display for MigrationError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            MigrationError::Io(e) => write!(f, "migration I/O error: {e}"),
            MigrationError::SqlFailed(v, e) => write!(f, "migration to v{v} failed (rolled back): {e}"),
            MigrationError::VersionAhead { db_version, app_version } => write!(
                f,
                "database schema v{db_version} is newer than this app (v{app_version}); \
                 upgrade the app to open this database"
            ),
        }
    }
}

/// Run all pending migrations on the database at `db_path`.
///
/// Returns `Ok(())` if no migrations were needed or all ran successfully.
/// Returns `Err(MigrationError)` if any migration failed; the database is unchanged.
pub fn run_migrations(db_path: &Path) -> Result<(), MigrationError> {
    let mut conn = Connection::open(db_path).map_err(|e| {
        MigrationError::Io(std::io::Error::new(std::io::ErrorKind::Other, e.to_string()))
    })?;

    // Enable WAL mode for crash safety.
    conn.execute_batch("PRAGMA journal_mode = WAL;").ok();
    conn.execute_batch("PRAGMA foreign_keys = ON;").ok();

    let current_version = get_user_version(&conn)?;

    if current_version == SCHEMA_VERSION_CURRENT {
        eprintln!("[migration] Schema is up to date (v{SCHEMA_VERSION_CURRENT}). No migrations needed.");
        return Ok(());
    }

    if current_version > SCHEMA_VERSION_CURRENT {
        return Err(MigrationError::VersionAhead {
            db_version: current_version,
            app_version: SCHEMA_VERSION_CURRENT,
        });
    }

    eprintln!(
        "[migration] Database is at v{current_version}, app requires v{SCHEMA_VERSION_CURRENT}. \
         Running {} migration(s).",
        SCHEMA_VERSION_CURRENT - current_version
    );

    // Create a backup before touching anything.
    create_backup(db_path, current_version)?;

    // Run each pending migration.
    for migration in MIGRATIONS.iter().filter(|m| m.to_version > current_version) {
        run_single_migration(&mut conn, migration, current_version)?;
    }

    eprintln!("[migration] All migrations complete. Schema is now v{SCHEMA_VERSION_CURRENT}.");
    Ok(())
}

/// Read `PRAGMA user_version` from an open connection.
fn get_user_version(conn: &Connection) -> Result<u32, MigrationError> {
    let ver: u32 = conn
        .query_row("PRAGMA user_version", [], |r| r.get(0))
        .map_err(|e| MigrationError::SqlFailed(0, e.to_string()))?;
    Ok(ver)
}

/// Set `PRAGMA user_version = N` on an open connection.
fn set_user_version(conn: &Connection, version: u32) -> SqlResult<()> {
    conn.execute_batch(&format!("PRAGMA user_version = {version}"))
}

/// Copy the database file to `<db_path>.bak.<old_version>.<timestamp>`.
fn create_backup(db_path: &Path, old_version: u32) -> Result<(), MigrationError> {
    let ts = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs();
    let bak_name = format!(
        "{}.bak.v{}.{}",
        db_path.file_name().and_then(|n| n.to_str()).unwrap_or("kairo.db"),
        old_version,
        ts
    );
    let bak_path = db_path.parent().unwrap_or(Path::new(".")).join(&bak_name);
    std::fs::copy(db_path, &bak_path).map_err(MigrationError::Io)?;
    eprintln!("[migration] Backup created at {bak_path:?}");
    Ok(())
}

/// Run a single migration in a transaction.
/// On failure: ROLLBACK → `Err(MigrationError::SqlFailed)`.
/// On success: COMMIT + set user_version.
fn run_single_migration(
    conn: &mut Connection,
    migration: &Migration,
    from_version: u32,
) -> Result<(), MigrationError> {
    eprintln!(
        "[migration] Applying migration → v{}: {}",
        migration.to_version, migration.description
    );

    // Use a savepoint so we can roll back just this migration if it fails.
    let tx = conn
        .transaction()
        .map_err(|e| MigrationError::SqlFailed(migration.to_version, e.to_string()))?;

    for &stmt in migration.statements {
        run_statement_idempotent(&tx, stmt).map_err(|e| {
            MigrationError::SqlFailed(migration.to_version, format!("SQL: {stmt}\nError: {e}"))
        })?;
    }

    // Record in audit log (best-effort — table may not exist in v0→v1 first run).
    let ts = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs() as i64;
    let _ = tx.execute(
        "INSERT INTO schema_migration_log (applied_at, from_version, to_version, description)
         VALUES (?1, ?2, ?3, ?4)",
        rusqlite::params![ts, from_version, migration.to_version, migration.description],
    );

    // Set user_version within the same transaction.
    tx.execute_batch(&format!("PRAGMA user_version = {}", migration.to_version))
        .map_err(|e| MigrationError::SqlFailed(migration.to_version, e.to_string()))?;

    tx.commit()
        .map_err(|e| MigrationError::SqlFailed(migration.to_version, e.to_string()))?;

    eprintln!("[migration] ✓ Migration to v{} succeeded.", migration.to_version);
    Ok(())
}

/// Execute a SQL statement, treating "duplicate column" and "table already exists"
/// errors as no-ops (idempotency for ALTER TABLE ADD COLUMN and CREATE TABLE IF NOT EXISTS).
fn run_statement_idempotent(conn: &Connection, sql: &str) -> SqlResult<()> {
    match conn.execute_batch(sql) {
        Ok(_) => Ok(()),
        Err(e) => {
            let msg = e.to_string().to_lowercase();
            // SQLite's error for adding an existing column: "duplicate column name"
            if msg.contains("duplicate column name") || msg.contains("already exists") {
                eprintln!("[migration/idempotent] Ignoring idempotent no-op: {msg}");
                Ok(())
            } else {
                Err(e)
            }
        }
    }
}

/// Check whether the Qdrant vector store meta matches the current embedding config.
///
/// Returns `true` if a rebuild is needed (dim or model changed).
/// Returns `false` if the store is up-to-date or doesn't exist yet.
pub fn qdrant_needs_rebuild(qdrant_dir: &Path, expected_dim: u32, expected_model: &str) -> bool {
    let meta_path = qdrant_dir.join("meta.json");
    if !meta_path.exists() {
        // New store — no rebuild needed, will be created fresh.
        return false;
    }

    match std::fs::read_to_string(&meta_path) {
        Ok(content) => {
            // Parse meta.json: {"embedding_dim": N, "model": "...", "schema_version": N}
            if let Ok(val) = serde_json::from_str::<serde_json::Value>(&content) {
                let dim_ok = val.get("embedding_dim")
                    .and_then(|v| v.as_u64())
                    .map(|d| d == expected_dim as u64)
                    .unwrap_or(false);
                let model_ok = val.get("model")
                    .and_then(|v| v.as_str())
                    .map(|m| m == expected_model)
                    .unwrap_or(false);
                !(dim_ok && model_ok)
            } else {
                eprintln!("[migration] Could not parse Qdrant meta.json — assuming rebuild needed.");
                true
            }
        }
        Err(e) => {
            eprintln!("[migration] Could not read Qdrant meta.json: {e} — assuming rebuild needed.");
            true
        }
    }
}

/// Write a new Qdrant meta.json after a successful rebuild.
pub fn write_qdrant_meta(qdrant_dir: &Path, dim: u32, model: &str, schema_version: u32) -> std::io::Result<()> {
    let meta = serde_json::json!({
        "embedding_dim": dim,
        "model": model,
        "schema_version": schema_version,
    });
    let meta_path = qdrant_dir.join("meta.json");
    std::fs::write(&meta_path, serde_json::to_string_pretty(&meta).unwrap())?;
    eprintln!("[migration] Qdrant meta.json written: dim={dim}, model={model}, schema_version={schema_version}");
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;
    use tempfile::TempDir;

    fn create_v0_db(dir: &Path) -> PathBuf {
        let db_path = dir.join("kairo.db");
        let conn = Connection::open(&db_path).unwrap();
        conn.execute_batch("
            CREATE TABLE documents (
                doc_id TEXT PRIMARY KEY,
                source_path TEXT,
                sha256 TEXT,
                page_count INTEGER,
                created_at INTEGER
            );
            CREATE TABLE pages (
                doc_id TEXT,
                page_index INTEGER,
                width_px INTEGER,
                height_px INTEGER,
                image_sha256 TEXT,
                PRIMARY KEY (doc_id, page_index)
            );
            CREATE TABLE extractions (
                id TEXT PRIMARY KEY,
                doc_id TEXT,
                field TEXT,
                value TEXT,
                confidence REAL,
                status TEXT
            );
            PRAGMA user_version = 0;
        ").unwrap();
        db_path
    }

    #[test]
    fn test_migration_v0_to_current() {
        let dir = TempDir::new().unwrap();
        let db_path = create_v0_db(dir.path());

        // Before migration: user_version = 0
        {
            let conn = Connection::open(&db_path).unwrap();
            let ver: u32 = conn.query_row("PRAGMA user_version", [], |r| r.get(0)).unwrap();
            assert_eq!(ver, 0, "Pre-migration version should be 0");
        }

        // Run migrations
        run_migrations(&db_path).expect("Migration should succeed");

        // After migration: user_version = SCHEMA_VERSION_CURRENT
        {
            let conn = Connection::open(&db_path).unwrap();
            let ver: u32 = conn.query_row("PRAGMA user_version", [], |r| r.get(0)).unwrap();
            assert_eq!(ver, SCHEMA_VERSION_CURRENT, "Post-migration version should be current");
        }

        // Backup file should exist
        let bak_files: Vec<_> = std::fs::read_dir(dir.path())
            .unwrap()
            .filter_map(|e| e.ok())
            .filter(|e| e.file_name().to_string_lossy().contains(".bak."))
            .collect();
        assert!(!bak_files.is_empty(), "Backup file must be created before migration");
    }

    #[test]
    fn test_migration_idempotent() {
        let dir = TempDir::new().unwrap();
        let db_path = create_v0_db(dir.path());

        // Run migrations twice — should not fail
        run_migrations(&db_path).expect("First migration should succeed");
        run_migrations(&db_path).expect("Second run (idempotent) should succeed");
    }

    #[test]
    fn test_backup_preserved_on_success() {
        let dir = TempDir::new().unwrap();
        let db_path = create_v0_db(dir.path());

        run_migrations(&db_path).expect("Migration should succeed");

        // Backup must still be present (not deleted on success).
        let bak_files: Vec<_> = std::fs::read_dir(dir.path())
            .unwrap()
            .filter_map(|e| e.ok())
            .filter(|e| e.file_name().to_string_lossy().contains(".bak."))
            .collect();
        assert!(!bak_files.is_empty(), "Backup must be preserved after successful migration");

        // Backup must be a valid SQLite file (readable).
        let bak_path = bak_files[0].path();
        let bak_conn = Connection::open(&bak_path).expect("Backup must be valid SQLite");
        let _ver: u32 = bak_conn
            .query_row("PRAGMA user_version", [], |r| r.get(0))
            .expect("Backup must be queryable");
    }

    #[test]
    fn test_version_ahead_returns_error() {
        let dir = TempDir::new().unwrap();
        let db_path = dir.path().join("future.db");
        let conn = Connection::open(&db_path).unwrap();
        // Set version much higher than current app version.
        conn.execute_batch("PRAGMA user_version = 999").unwrap();
        drop(conn);

        let result = run_migrations(&db_path);
        assert!(
            matches!(result, Err(MigrationError::VersionAhead { .. })),
            "Should error on version-ahead database"
        );
    }

    #[test]
    fn test_qdrant_needs_rebuild_no_meta() {
        let dir = TempDir::new().unwrap();
        // No meta.json → no rebuild needed (fresh store)
        assert!(!qdrant_needs_rebuild(dir.path(), 256, "hash-256"));
    }

    #[test]
    fn test_qdrant_needs_rebuild_stale_model() {
        let dir = TempDir::new().unwrap();
        let meta = serde_json::json!({
            "embedding_dim": 384,
            "model": "sentence-transformers/all-MiniLM-L6-v2",
            "schema_version": 1,
        });
        std::fs::write(dir.path().join("meta.json"), meta.to_string()).unwrap();

        // Different dim and model → rebuild needed
        assert!(qdrant_needs_rebuild(dir.path(), 256, "hash-256"));
    }

    #[test]
    fn test_qdrant_no_rebuild_when_matching() {
        let dir = TempDir::new().unwrap();
        let meta = serde_json::json!({
            "embedding_dim": 256,
            "model": "hash-256",
            "schema_version": 2,
        });
        std::fs::write(dir.path().join("meta.json"), meta.to_string()).unwrap();

        assert!(!qdrant_needs_rebuild(dir.path(), 256, "hash-256"));
    }
}
